Override capability device_id in update request

w1ll1am23 opened the pull request:
So, in my last PR when I added the sensor capability to the device id it broke the update sort of. So in order to keep a unique ID and still update properly the device_id needs overridden in the update request. Thankfully you already added in the ability to override the ID. 

Going to look into the garage issue later today or tomorrow. Figured we should wait for that fix before updating the version.

Commit:
Override capability device_id in update request

bradsk88 commented:
Good find. Since this is a regression, I'd like you to add a test or two to keep it from happening again.

w1ll1am23 commented:
Sounds good, I'll try to work on that today or tomorrow as well.
On Mar 13, 2016 1:43 PM, "Brad Johnson" notifications@github.com wrote:

> Good find. Since this is a regression, I'd like you to add a test or two
> to keep it from happening again.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/bradsk88/python-wink/pull/24#issuecomment-196011957.

w1ll1am23 commented:
I moved all of the tests into one file, and I also moved all of the api responses into their own files. Not sure if this is how you want to do this or not. Also I am not sure if the check for the device_id is "good enough." Let me know what you think.

bradsk88 commented on src/pywink/test/init_test.py:
```diff
@@ -45,771 +55,256 @@ def test_should_only_send_color_xy_if_both_color_xy_and_color_temperature_are_gi
+        with open('{}/api_responses/power_strip.json'.format(os.path.dirname(__file__))) as powerstrip_file:
+            response_dict = json.load(powerstrip_file)
+        power_strip = response_dict.get('data')
+        outlets = power_strip[0].get('outlets')
```
There is no assertion in this test

bradsk88 commented on src/pywink/test/init_test.py:
```diff
@@ -45,771 +55,256 @@ def test_should_only_send_color_xy_if_both_color_xy_and_color_temperature_are_gi
+        garage_door = response_dict.get('data')[0]
+        wink_garage_door = WinkGarageDoor(garage_door, self.api_interface)
+        device_id = wink_garage_door.device_id()
+        self.assertRegex(device_id, "^[0-9]{4,6}$")
```
I think what you're actually asserting here is that the device ID **starts** with a number

w1ll1am23 replied:
This should be only a number ^ = beginning of line and $ = end of line. Now on the sensor I am only checking that it starts with a number since it also returns the capability.

bradsk88 replied:
Right sorry.  I confused the two code blocks.

bradsk88 commented on src/pywink/test/init_test.py:
```diff
@@ -45,771 +55,256 @@ def test_should_only_send_color_xy_if_both_color_xy_and_color_temperature_are_gi
+        expected_temperature = 5
+        self.assertEquals(expected_temperature, temp_sensor.temperature_float())
+
+    def test_device_id_should_be_number(self):
```
`should_start_with_a_number`

bradsk88 commented:
These tests are ok but, if you undo your changes to `update_state` (effectively putting this bug "back"), these tests continue to pass.

A regression test should fail when the bug is present and pass after it's fixed.

Example incoming.

w1ll1am23 commented:
Ahhh that makes sense, didn't think about them failing. I will add another test to verify that the URL request ends in a number. I think that should work?

bradsk88 commented:
Example of a regression test:

```
    def test_update_state_should_call_get_state_endpoint_with_capability_removed_from_id(self):
        expected_id = 'arbitrary ID'
        fake_api_interface = mock.MagicMock()
        sensor = WinkTemperatureSensor({
            'sensor_pod_id': expected_id
        }, fake_api_interface)

        sensor.UNIT = 'DEG'  # mock doesn't like unicode

        sensor.update_state()
        fake_api_interface.get_device_state.assert_called_once_with(sensor, expected_id)

```

Note that rather than adding assertions for what URLs were called, I can leverage the fact that we use dependency injection for the API interface and inject a mock object for assertions.

Assuming you know how MagicMock works (a reasonable assumption because it's a core python library), this allows the test to be very specific and simplistic.

It also keeps us un-tethered from the current implementation of the API interfact so, if we were to do something absolutely crazy like offering LAN support, we won't have to even touch this test.

w1ll1am23 commented:
> Example of a regression test:
> 
> ```
> def test_should_call_get_state_endpoint_with_capability_removed_from_id(self):
>     expected_id = 'arbitrary ID'
>     fake_api_interface = mock.MagicMock()
>     sensor = WinkTemperatureSensor({
>         'sensor_pod_id': expected_id
>     }, fake_api_interface)
> 
>     sensor.UNIT = 'DEG'  # mock doesn't like unicode
> 
>     sensor.update_state()
>     fake_api_interface.get_device_state.assert_called_once_with(sensor, expected_id)
> ```

Nice, makes sense. So is it enough to have this check with just one of the sensors like you are doing? since they all use the same method?

bradsk88 commented:
@w1ll1am23 Ha!  I was worried you might ask that.

Under the most strict adherence to testing practises: no, it's not ok because we're actually testing the behaviour of _WinkCapabilitySensor.  It's possible that WinkTemperatureSensor could again override this function which would detach this test from its original intent.

The correct solution would be to define a new class, in our test file, specifically for this test.

```
class TestWinkCapabilitySensor(_WinkCapabilitySensor):

  self.unit = 'ANYTHING'
  self.capability = 'SOMETHING'
```

and then you would use that class in place of `WinkTemperatureSensor` in my above example.

bradsk88 commented:
Looks good to me.
Update the minor version number and I'll merge it in.
